src/world_model/nav2_costmap_world_model/src/costmap_world_model.cpp

```diff
@@ -27,32 +28,78 @@ namespace nav2_costmap_world_model
+  // TODO(bpwilcox): replace manual footprint to layered_costmap with parameter or from nav2_robot
+  setFootprint(0, 0);
+  layered_costmap_->updateMap(0, 0, 0);
+
 
   auto costmap_service_callback = [this](
-    const std::shared_ptr<rmw_request_id_t>/*request_header*/,
+    const std::shared_ptr<rmw_request_id_t> request_header,
     const std::shared_ptr<nav2_world_model_msgs::srv::GetCostmap::Request> request,
     const std::shared_ptr<nav2_world_model_msgs::srv::GetCostmap::Response> response) -> void
     {
       RCLCPP_INFO(
         this->get_logger(), "CostmapWorldModel::CostmapWorldModel:Incoming costmap request");
-      response->map = costmap_->getCostmap(request->specs);
+      costmap_callback(request_header, request, response);
     };
 
   // Create a service that will use the callback function to handle requests.
   costmapServer_ = create_service<nav2_world_model_msgs::srv::GetCostmap>(name + "_GetCostmap",
       costmap_service_callback);
+}
+
+void CostmapWorldModel::costmap_callback(
+  const std::shared_ptr<rmw_request_id_t>/*request_header*/,
+  const std::shared_ptr<nav2_world_model_msgs::srv::GetCostmap::Request>/*request*/,
+  const std::shared_ptr<nav2_world_model_msgs::srv::GetCostmap::Response> response)
+{
+  costmap_2d::Costmap2D * costmap = layered_costmap_->getCostmap();
+  rclcpp::Clock clock;
 
-  // Get the current map from the map server
-  //
-  // TODO(mjeronimo): Instead of using a service call, the map server should push any
-  // map updates using a latched topic. Unfortunately, no latched topics yet in ROS2
-  map_client_.waitForService(std::chrono::seconds(2));
+  response->map.metadata.size_x = costmap->getSizeInCellsX();
+  response->map.metadata.size_y = costmap->getSizeInCellsY();
+  response->map.metadata.resolution = costmap->getResolution();
+  response->map.metadata.layer = "Master";
+  response->map.metadata.map_load_time = now();
+  response->map.metadata.update_time = now();
 
-  auto request = std::make_shared<nav2_tasks::MapServiceClient::MapServiceRequest>();
-  auto response = map_client_.invoke(request);
+  tf2::Quaternion quaternion;
+  quaternion.setRPY(0.0, 0.0, 0.0);  // set roll, pitch, yaw
+  response->map.metadata.origin.position.x = costmap->getOriginX();
+  response->map.metadata.origin.position.y = costmap->getOriginY();
+  response->map.metadata.origin.position.z = 0.0;
+  response->map.metadata.origin.orientation = tf2::toMsg(quaternion);
 
-  costmap_->setStaticMap(response->map);
+  response->map.header.stamp = now();
+  response->map.header.frame_id = "map";
+
+  unsigned char * data = costmap->getCharMap();
+  auto data_length = response->map.metadata.size_x * response->map.metadata.size_y;
+  response->map.data.resize(data_length);
+  response->map.data.assign(data, data + data_length);
+}
+
+void CostmapWorldModel::setFootprint(double length, double width)
+{
+  std::vector<geometry_msgs::msg::Point> polygon;
+  geometry_msgs::msg::Point p;
+  p.x = width;
```

Comment:
You end up making the footprint 2 * width and length here. Seems the footprint should be either.
p(w/2, l/2), p(w/2, -l/2), p(-w/2, -l/2), p(-w/2, -l/2)
or
p(0,0), p(w, 0), p(0,l),p(w,l)?

Reply:
I stripped this code straight from the test code in the original costmap_2d, so you may be right on a more sensible footprint. It really is just a stubbed value for now, though.

Reply:
I agree with Carl, even if you got the code from somewhere else, I think you should fix it.